File: C++/프로그래머스/2/42839. 소수 찾기/소수 찾기.cpp
#include <string>
#include <vector>
#define MAX 10000000
using namespace std;
enum Prime {
    Prime,
    NonPrime,
};
void get_primes(vector<bool>& primes) {
    primes[0] = NonPrime;
    primes[1] = NonPrime;
    for (int i = 2; i * i < MAX; ++i) {
        if (primes[i]) {
            continue;
        }
        for (int j = i * i; j < MAX; j += i) {
            primes[j] = NonPrime;
        }
    }
}
void back_tracking(const vector<bool>& primes, string& numbers, vector<bool>& visit,int& ret, int& answer) {
    for (int i = 0; i < numbers.length(); ++i) {
        if (numbers[i] == '-') {
            continue;
        }
        ret = ret * 10 + (numbers[i] - '0');
        if (primes[ret] == Prime && !visit[ret]) {
            visit[ret] = true;
            answer += 1;
        }
        char tmp = numbers[i];
        numbers[i] = '-';
        back_tracking(primes, numbers, visit, ret, answer);
        numbers[i] = tmp;
        ret /= 10;
    }
}
int solution(string numbers) {
    int answer = 0;
    int ret = 0;
    vector<bool> primes(MAX);
    vector<bool> visit(MAX);
    get_primes(primes);
    back_tracking(primes, numbers, visit, ret, answer);
    return answer;
}

File: Java/프로그래머스/2/42839. 소수 찾기/소수 찾기.java
class Solution {
    public static final int MAX = 10000000;
    public static boolean[] nonPrime = new boolean[MAX];
    public static boolean[] check = new boolean[MAX];
    public static int maxLen;
    public static int count = 0;
    public static char[] cArr;
    
    public void recur(int idx, int num) {
        if (idx == maxLen) return;
        for (int i = 0; i < maxLen; i++) {
            if (cArr[i] == '-') continue;
            
            char tmp = cArr[i];
            cArr[i] = '-';
            num = num * 10 + ((int)tmp - '0');
            if (!nonPrime[num] && !check[num]) {
                count++;
                check[num] = true;
            }
            recur(idx + 1, num);
            cArr[i] = tmp;
            num /= 10;
        }
    }
    
    
    
    public int solution(String numbers) {
        cArr = numbers.toCharArray();
        maxLen = numbers.length();
        
        nonPrime[0] = true;
        nonPrime[1] = true;
        for (int i = 2; i * i < MAX;) {
            for (int n = i * i; n < MAX; n += i)
                nonPrime[n] = true;
            int next = i + 1;
            while (nonPrime[next])
                next++;
            i = next;
        }
        recur(0, 0);
        return count;
    }
    
}